### PR TITLE
Initial gcc toolchain configuration for the lowRISC compiler.

### PR DESCRIPTION
**Pull request by cfrantz**
Configure lowrisc_toolchain_rv32imc for use by the OpenTitan project.

### CODE REVIEW

**Comment by cfrantz** [comment]
Things to discuss:
1) What is the proper name in `constraints/cpu/BUILD`: riscv32 or rv32imc?  Based on the [bazelbuild/platforms](https://github.com/bazelbuild/platforms/blob/main/cpu/BUILD#L123) repo, I feel like `riscv32` is the name of the cpu and `rv32imc` is a more specific architecture-and-features descriptor.

2) Need to do better on compiler flag selections in `toolchains/features/common/impl/gcc.bzl` and `toolchains/features/embedded/impl/gcc.bzl`.  I think these flag selections should be data driven and the flags themselves should be moved into the platform or device configurations: `platforms/BUILD` and/or `devices/riscv32/riscv32.bzl`.  This is a bit of a bigger structural change and will probably require some consultation with the upstream maintainers.

File: toolchains/features/common/impl/gcc.bzl [hunk 1]
```diff
@@ -49,31 +49,56 @@ def GccIncludeFeature(include_paths):
...
-                            # Set the fpu available on the chip
-                        ],
-                    ),
-                ],
-            ),
-        ],
-    )
+    if architecture == 'rv32imc':
+        # TODO(cfrantz): why do I have to do this?  Seems the lowRISC gcc
+        # doesn't support `-mfpu`, `-mfloat-abi` or `-mlittle-endian`.
```
Comment (mundaym) [inline toolchains/features/common/impl/gcc.bzl:54]: `-mlittle-endian` should work (at least it should if we update gcc). AFAICT the other flags are arm specific and aren't defined for other targets. The flag definitions should probably be moved into arch-specific code.

File: constraints/cpu/BUILD [hunk 2]
```diff
@@ -15,3 +15,12 @@ constraint_value(
     constraint_setting = "@platforms//cpu:cpu",
     visibility = ["//visibility:public"],
 )
+
+# TODO(cfrantz): Upstream to https://github.com/bazelbuild/platforms
+constraint_value(
+    # Based on the current riscv64 consrtaint in platforms, I wonder
+    # if this should be `riscv32` or `rv32imc`.
+    name = "rv32imc",
```
Comment (mundaym) [inline constraints/cpu/BUILD]: I'm leaning towards keeping `rv32imc`. It looks like the generic `arm` target is slated for removal and there are a bunch of more specific 32-bit arm targets. `riscv64` is probably geared towards the non-embedded use case.

Comment (cfrantz) [reply constraints/cpu/BUILD]: Based on https://github.com/bazelbuild/platforms/pull/30, `riscv32` is the appropriate high-level cpu descriptor.  In fact, the change has already been applied by Google's copybara automation to Google internal copy of the `platforms` repo.

**Comment by cfrantz** [comment]
To be consistent with bazelbuild/platforms#30, I've renamed the cpu to `riscv32`.

**Comment by cfrantz** [comment]
I think we should commit this change as-is to move the bazel work foward.

I have opened https://github.com/silvergasp/bazel-embedded/issues/37 with upstream to try to get feedback on the proper approach to integrating compiler/feature flags into a "user-serviceable" part of the bazel-embedded repo.  Until we have that understanding, its fine to live with the `if architecture == 'riscv32'` hacks in `toolchains/features/{common, embedded}/impl/gcc.bzl`.

**Comment by mundaym** [comment]
> I think we should commit this change as-is to move the bazel work foward.

SGTM. Feel free to commit this.

File: toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    ))
+    include_paths = include_tools.ProccessResponse(response.stderr)
+    include_flags = ["-isystem" + path for path in include_paths]
+    include_bazel_template_input = include_tools.CommandLineToTemplateString(include_flags)
+    include_paths_template_input = include_tools.CommandLineToTemplateString(include_paths)
+    repository_ctx.file(
+        "defs.bzl",
+        """
+SYSTEM_INCLUDE_COMMAND_LINE = {}
+SYSTEM_INCLUDE_PATHS= {}
```
Comment (mcy) [inline toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl]: Space after the = sign? =P

File: toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+""".format(include_bazel_template_input, include_paths_template_input),
+    )
+
+    repository_ctx.file(
+        "BUILD",
+        """
+filegroup(
+    name = "all",
+    srcs = glob(["**/*"],exclude=["**/*.html","**/*.pdf"]),
```
Comment (mcy) [inline toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl]: Missing space after the , too

**Review by mcy — COMMENTED** [review]
Just a couple of format nits. I'm going to assume Michael's LGTM is good enough. =)

File: toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl (hunk 3 above)
Comment (cfrantz) [reply toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl]: Fixed.  These two formatting errors also exist in the file I copied this from (`gcc_arm_none_repository.bzl`)

File: toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl (hunk 4 above)
Comment (cfrantz) [reply toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl]: Fixed.

**Review by mundaym — APPROVED** [review]

**Pull request merged**